lib/src/generator_frontend.dart

```diff
@@ -0,0 +1,377 @@
+import 'package:dartdoc/src/model/model.dart';
+import 'package:dartdoc/src/model_utils.dart';
+import 'package:dartdoc/src/warnings.dart';
+import 'package:path/path.dart' as path;
+
+typedef FileWriter = File Function(String filePath, Object content,
+    {bool allowOverwrite, Warnable element});
+
+/// [Generator] that delegates rendering to a [GeneratorBackend] and delegates
+/// file creation to a [FileWriter].
+class GeneratorFrontEnd implements Generator {
+  final GeneratorBackend _generatorBackend;
+  final FileWriter _writer;
+
+  // Used in write(). This being not null signals the generator is active.
+  String _outputDirectory;
+
+  final StreamController<void> _onFileCreated = StreamController(sync: true);
+
+  @override
+  Stream<void> get onFileCreated => _onFileCreated.stream;
+
+  @override
+  final Map<String, Warnable> writtenFiles = {};
+
+  GeneratorFrontEnd(this._generatorBackend, this._writer);
+
+  // Implement FileWriter so we can wrap the one given to us.
+  File write(String filePath, Object content,
+      {bool allowOverwrite, Warnable element}) {
+    assert(_outputDirectory != null);
+    // Replace '/' separators with proper separators for the platform.
+    String outFile = path.joinAll(filePath.split('/'));
+    outFile = path.join(_outputDirectory, outFile);
+    File file = _writer(outFile, content,
+        allowOverwrite: allowOverwrite, element: element);
+    writtenFiles[outFile] = element;
+    _onFileCreated.add(file);
+    return file;
+  }
+
+  @override
+  Future generate(PackageGraph packageGraph, String outputDirectory) async {
+    _outputDirectory = outputDirectory;
+    try {
+      List<Indexable> indexElements = <Indexable>[];
+      _generateDocs(packageGraph, indexElements);
+      await _generatorBackend.generateAdditionalFiles(write, packageGraph);
+
+      List<Categorization> categories = indexElements
+          .where((e) => e is Categorization && e.hasCategorization)
+          .map((e) => e as Categorization)
+          .toList();
+      _generatorBackend.generateCategoryJson(write, categories);
+      _generatorBackend.generateSearchIndex(write, indexElements);
+    } finally {
+      _outputDirectory = null;
+    }
+  }
+
+  // Traverses the package graph and collects elements for the search index.
+  void _generateDocs(
+      PackageGraph packageGraph, List<Indexable> indexAccumulator) {
+    if (packageGraph == null) return;
+
+    logInfo('documenting ${packageGraph.defaultPackage.name}');
+    _generatorBackend.generatePackage(
+        write, packageGraph, packageGraph.defaultPackage);
+
+    for (var package in packageGraph.localPackages) {
+      for (var category in filterNonDocumented(package.categories)) {
+        logInfo('Generating docs for category ${category.name} from '
+            '${category.package.fullyQualifiedName}...');
+        indexAccumulator.add(category);
+        _generatorBackend.generateCategory(write, packageGraph, category);
```

Comment:
Yes, this approach is an improvement.  Eventually it seems like we should be able to refactor and use a visitor pattern but I think this is definitely good enough for now.